File: src/main/java/club/deneb/client/features/Module.java
package club.deneb.client.features;

import club.deneb.client.client.GuiManager;
import club.deneb.client.utils.clazz.Button;
import club.deneb.client.value.ButtonValue;
import club.deneb.client.value.StringMode;
import club.deneb.client.value.Value;
import org.lwjgl.input.Keyboard;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by B_312 on 01/10/21
 */
public class Module extends AbstractModule {

    Value<String> visible_value;
    Value<Button> resetConfig;

    public boolean isShownOnArray(){
        return this.visible_value.toggled("ON");
    }

    public Module(){
        this.name = getAnnotation().name();
        this.category = getAnnotation().category();
        this.description = getAnnotation().description();
        this.keyCode = getAnnotation().keyCode();
        this.getValues().add(visible_value = new StringMode("Visible",getAnnotation().visible() ? "ON" : "OFF" , listOf("ON","OFF")).v(GuiManager.INSTANCE::getVisibleButton));
        this.getValues().add(resetConfig = new ButtonValue("LoadDefault", new Button().setBind(this::reset)).des("Click here to reset this module").v(GuiManager.INSTANCE::getResetButton));
        this.isHUD = false;
        this.onInit();
    }

    public void onInit(){}

    @Retention(RetentionPolicy.RUNTIME)
    public @interface Info {
        String name();
        String description() default "";
        int keyCode() default Keyboard.KEY_NONE;
        Category category();
        boolean visible() default true;
    }

    private Info getAnnotation() {
        if (getClass().isAnnotationPresent(Info.class)) {
            return getClass().getAnnotation(Info.class);
        }
        throw new IllegalStateException("No Annotation on class " + this.getClass().getCanonicalName() + "!");
    }

}

File: src/main/java/club/deneb/client/features/Module.kt
package club.deneb.client.features

import club.deneb.client.client.GuiManager
import club.deneb.client.utils.clazz.Button
import club.deneb.client.utils.clazz.VoidContainer
import club.deneb.client.value.ButtonValue
import club.deneb.client.value.StringMode
import club.deneb.client.value.Value
import org.lwjgl.input.Keyboard

open class Module : AbstractModule() {

    private var visibleValue: Value<String>
    private var resetConfig: Value<Button>

    val isShownOnArray: Boolean
        get() = visibleValue.toggled("ON")

    @kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
    annotation class Info(
        val name: String,
        val description: String = "",
        val keyCode: Int = Keyboard.KEY_NONE,
        val category: Category,
        val visible: Boolean = true
    )

    private val annotation: Info
        get() {
            if (javaClass.isAnnotationPresent(Info::class.java)) {
                return javaClass.getAnnotation(Info::class.java)
            }
            throw IllegalStateException("No Annotation on class " + this.javaClass.canonicalName + "!")
        }

    init {
        name = annotation.name
        this.category = annotation.category
        description = annotation.description
        keyCode = annotation.keyCode
        values.add(
            StringMode("Visible", if (annotation.visible) "ON" else "OFF", listOf("ON", "OFF")).v(GuiManager::visibleButton).also { visibleValue = it })
        values.add(
            ButtonValue("LoadDefault", Button().setBind(object : VoidContainer {
                override fun invoke() {
                    reset()
                }
            })).des("Click here to reset this module").v(GuiManager::resetButton).also { resetConfig = it })
        isHUD = false
    }
}